compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -507,8 +509,45 @@ macro_rules! define_queries {
             })*
         }
 
+        mod query_structs {
+            use $crate::plumbing::QueryStruct;
+            use rustc_query_system::query::QueryDescription;
+            $(
+            pub(super) const fn $name<'tcx>() -> QueryStruct<'tcx> { QueryStruct {
+                try_collect_active_jobs: |tcx, qmap| {
+                    let make_query = |tcx, key| {
+                        let kind = rustc_middle::dep_graph::DepKind::$name;
+                        let name = stringify!($name);
+                        $crate::plumbing::create_query_frame(tcx, super::queries::$name::describe, key, kind, name)
+                    };
+                    tcx.queries.$name.try_collect_active_jobs(
+                        tcx,
+                        make_query,
+                        qmap,
+                    )
+                },
+                alloc_self_profile_query_strings: |tcx, string_cache| {
+                    $crate::profiling_support::alloc_self_profile_query_strings_for_query_cache(
+                        tcx,
+                        stringify!($name),
+                        &tcx.query_caches.$name,
+                        string_cache,
+                    )
+                },
+                encode_query_results: expand_if_cached!([$($modifiers)*], |tcx, encoder, query_result_index|
+                    $crate::on_disk_cache::encode_query_results::<_, super::queries::$name<'_>>(tcx, encoder, query_result_index)
+                ),
+            }})*
+        }
+
         pub fn query_callbacks<'tcx>(arena: &'tcx Arena<'tcx>) -> &'tcx [DepKindStruct<'tcx>] {
-            arena.alloc_from_iter(make_dep_kind_array!(query_callbacks))
+            arena.alloc_from_iter(make_dep_kind_array!(query_callbacks with_builtin_nodes))
+        }
+
+        fn for_each_query_struct<'tcx>(mut cb: impl FnMut(&$crate::plumbing::QueryStruct<'tcx>)) {
+            for query in make_dep_kind_array!(query_structs) {
+                cb(&query);
```

Comment:
I'm not super happy with this setup; the callback is awkward and `make_dep_kind_array` is being stored into a local variable, which I worry will blow up the call stack. But I'll wait until we get the results back from the perf run to try and modify it more.

Originally I tried:
- Storing `make_dep_kind_array` into a static, which didn't work because the function pointers are invariant, so I can't convert them from `fn(TyCtxt<'static>)` to `fn(TyCtxt<'tcx>)`
- Storing `make_dep_kind_array` into TyCtxt in the same way as `query_callbacks`. That didn't work because `QueryStruct` and friends aren't defined until rustc_query_impl, so there's no way to name them.

Reply:
What about having this array in `Queries<'tcx>`?

Reply:
I don't understand how we would use that. `Queries` isn't available in `TyCtxt`; that's ok for `alloc_self_profile_query_strings` because it's only called in `rustc_interface`, but `OnDiskCache::serialize` only has `OnDiskCache` and `TyCtxt` available.

Reply:
Oh, I see, it's stored as `&dyn QueryEngine`. Hmm, I'm still not sure how this would work though, since we can't name `QueryStruct` in the QueryEngine trait.

Reply:
I tried downcasting (using `dyn Any`), but that doesn't work because `Queries<'tcx>` doesn't implement `'static`.

Reply:
You have 3 callers:
- `QueryCtxt::encode_query_results`, which has access to a `Queries<'tcx>`;
- `Queries::try_collect_active_jobs`, which also has such access,
- `alloc_self_profile_query_strings` which could use `QueryCtxt::from_tcx(tcx)` to get it, like you said.

The downcasting through `Any` is what `QueryCtxt::from_tcx` does.

Reply:
That worked :) thanks!